Sources/AuthFoundation/User Management/Credential.swift

```diff
@@ -301,55 +337,127 @@ public final class Credential: Equatable, OAuth2ClientDelegate {
+                }
+
+                let result: Token.Metadata
+
+                let id = id
+                if let coordinator,
+                   let metadata = withAsyncGroup({
+                       try? await coordinator.tokenStorage.metadata(for: id)
+                   })
+                {
+                    result = metadata
+                } else {
+                    result = Token.Metadata(id: id)
+                }
+
+                _metadata = result
+                return result
+            }
         }
-        
-        return Token.Metadata(id: id)
-    }()
-    
-    internal private(set) var automaticRefreshTimer: DispatchSourceTimer?
+        set {
+            lock.withLock {
+                _metadata = newValue
+            }
+        }
+    }
+
+    nonisolated(unsafe) private var _automaticRefresh: Bool = false
+    nonisolated(unsafe) private(set) var _automaticRefreshTimer: DispatchSourceTimer?
+    nonisolated(unsafe) private var _automaticRefreshTask: Task<Void, Never>?
+
     private func startAutomaticRefresh() {
-        guard let timerSource = createAutomaticRefreshTimer() else { return }
+        guard let expiresAt = _token.expiresAt
+        else {
+            return
+        }
+        let graceInterval = Credential.refreshGraceInterval
+        let timeOffset = max(0.0, expiresAt.timeIntervalSinceNow - Date.nowCoordinated.timeIntervalSinceNow - graceInterval)
+        let repeatInterval = min(_token.expiresIn - graceInterval, _token.expiresIn)
+
+        _automaticRefreshTask = Task(priority: .medium) {
+            if timeOffset > 0 {
+                do {
+                    try await Task.sleep(nanoseconds: UInt64(timeOffset) * _CredentialAutomaticRefreshTimeIntervalToNanoseconds)
+                } catch is CancellationError {
+                    lock.withLock {
+                        _automaticRefreshTask = nil
+                    }
+                    return
+                } catch {
+                    print("Error during automatic refresh: \(error)")
+                }
+            }
 
-        automaticRefreshTimer?.cancel()
-        automaticRefreshTimer = timerSource
-        timerSource.resume()
+            repeat {
+                do {
+                    try await refreshIfNeeded()
+                    try await Task.sleep(nanoseconds: UInt64(repeatInterval) * _CredentialAutomaticRefreshTimeIntervalToNanoseconds)
+                } catch is CancellationError {
+                    lock.withLock {
+                        _automaticRefreshTask = nil
+                    }
+                } catch {
+                    print("Error during automatic refresh: \(error)")
+                }
+            } while !Task.isCancelled
+        }
     }
-    
-    func stopAutomaticRefresh() {
-        automaticRefreshTimer?.cancel()
-        automaticRefreshTimer = nil
+
+    private func stopAutomaticRefresh() {
+        _automaticRefreshTask?.cancel()
```

Comment:
There is no synchronization with `startAutomaticRefresh` function. `_automaticRefreshTask ` can be overwritten with new task, so previous will never be cancelled

Reply:
These functions are private, and are only called from the `automaticRefresh` setter.  They're locked at the setter level, so special care is taken within these functions to ensure that no inadvertent accesses are made to public nonisolated properties which use the locks (e.g. it uses `_token` instead of `token`).

Additionally, when the task needs to be cancelled in response to an async operation within the Task itself, it re-locks its access to `self` before niling out the refresh task.